### ISSUE TITLE
libstdc++-6.dll causes segfault in app compiled with Clang

### ISSUE THREAD

**Issue by ghost** [Alexpux/MINGW-packages#735]
I know it's a wrong place for resolving problems in GNU C++ standard library, but currently crash happens only when application is compiled with Clang from MSYS2:

``` cpp
#include <unordered_map>
#include <string>

int main()
{
  std::unordered_map<std::string, int> test;
  test["cool"] = 5;
}
```

```
Program received signal SIGSEGV, Segmentation fault.
0x000000006fc9a07b in ?? () from C:\msys64\mingw64\bin\libstdc++-6.dll
(gdb) bt
#0  0x000000006fc9a07b in ?? () from C:\msys64\mingw64\bin\libstdc++-6.dll
#1  0x00000000004019c6 in _M_insert_unique_node ()
    at C:\msys64\mingw64\bin/../include/c++/5.2.0\bits/hashtable.h:1594
#2  0x0000000000401962 in operator[] ()
    at C:\msys64\mingw64\bin/../include/c++/5.2.0\bits/hashtable_policy.h:625
#3  0x0000000000401622 in operator[] ()
    at C:\msys64\mingw64\bin/../include/c++/5.2.0\bits/unordered_map.h:672
#4  main () at ../src\main.cpp:7
```

Any ideas / suggestions how to debug it further? Is there any debug symbols available anywhere for libstdc++?

Thanks!

**Comment by martell** [comment]
Hi @procedural

Can you tell me the clang version?
And the link flags you passed ? 

Possibly related but doubtful #722

Also can you create an issue similar to this 
https://llvm.org/bugs/show_bug.cgi?id=24398

with the above test case and assign it to me

Cheers 
Martell :)

**Comment by procedural** [comment]
Hi Martell!

>Can you tell me the clang version?

```
clang version 3.6.2 (tags/RELEASE_362/final)
Target: x86_64-w64-windows-gnu
Thread model: posix
```

>And the link flags you passed ?

Didn't passed anything, can be reproduced with these command line:

```
clang++ -std=c++11 main.cpp
a.exe
```

>Possibly related but doubtful #722

There's no backtraces, hard to tell.

>Also can you create an issue similar to this 
https://llvm.org/bugs/show_bug.cgi?id=24398
with the above test case and assign it to me

Sure! Thanks for pointing!

**Comment by procedural** [comment]
https://llvm.org/bugs/show_bug.cgi?id=24408

Don't know how to assign you in that bug-tracker, though :)

**Comment by martell** [comment]
@procedural Can you check again by building the latest clang-svn
I don't have my windows box atm.

I have a feeling it was fixed in 
http://reviews.llvm.org/rL245084
and also in i686 in
http://reviews.llvm.org/rL245459

**Comment by procedural** [comment]
>by building the latest clang-svn

You mean installing the latest clang-svn from MSYS2? Building Clang from sources is a slow and painful process for me...

**Comment by martell** [comment]
You can probably download a RC build of 3.7 for windows from llvm.org and set -target x86_64-windows-gnu to test.
It appears both this and the other bug report are older issues that were fixed long ago on HEAD before 3.7 were branched

**Comment by procedural** [comment]
>It appears both this and the other bug report are older issues that were fixed long ago on HEAD before 3.7 were branched

Then you can close it and if it won't be the case we always can reopen the bug.

Thanks for pointing that out! I'll give 3.7 a try...

**Comment by awson** [comment]
Freshly built SVN clang also produces segfaulting exe, so no fixes yet.

**Comment by martell** [comment]
I believe this is related to #722 and has been fixed by
http://reviews.llvm.org/rL251567
please retest on HEAD

**Comment by awson** [comment]
In fact I've fixed this myself quite a long ago, and the problem is different.

Sorry, I've turned to be too hesitant to give the patch back to the community because after reading llvm/clang-related mailing lists I understood that the community is, perhaps, not so easy to interact with.

Anyway, the bug is so egregious that I wonder how it survived for so long time.

Look:
https://github.com/llvm-mirror/clang/blob/f6a5ffd2cec68e89ddfb67b3c078c6c86a516a76/lib/CodeGen/TargetInfo.cpp#L3336

This is completely and utterly wrong. What is interesting, this existed ever since the 64-bit windows port was created. I suspect that was related either to some peculiarity in old gcc 4.6 or to plain mistake. Even clang documentation mentions this as a potential problem.

I've checked all gcc versions starting at 4.8 and all of them conform to standard Windows 64-bir ABI.

Thus, the fix is simply to remove lines 3336-3339 from the code.

I believe it would be easier to you (being an active llvm/clang community member) to land this change into upstream clang, thanks.
